crates/walrus-service/src/node/consistency_check.rs

```diff
@@ -72,26 +117,96 @@ fn get_epoch_bucket(epoch: Epoch) -> String {
+        Ok(true) => *total_fully_stored += 1,
+        Ok(false) => {}
+        Err(error) => {
+            *existence_check_error += 1;
+            tracing::warn!(?error, blob_id=%blob_id, "error when checking sliver data existence");
+        }
+    }
+}
+
+/// Compose the digest of the blob list returned by the iterator, and check the existence of the
+/// sliver data of the blobs that are fully synced.
+///
 /// `scan_counter` keeps track of the number of blobs scanned.
-fn compose_blob_list_digest<B: AsRef<[u8]>, T: CertifiedBlobInfoApi>(
-    blob_info_iter: BlobInfoIter<
-        B,
-        T,
-        impl Iterator<Item = Result<(B, T), TypedStoreError>> + Send + ?Sized,
-    >,
+fn compose_blob_list_digest_and_check_sliver_data_existence(
+    node: &StorageNodeInner,
+    blob_info_iter: BlobInfoIterator<'_>,
     epoch: Epoch,
     scan_counter: &IntCounterVec,
-) -> Result<u64, TypedStoreError> {
+    blobs_not_yet_fully_synced: &HashSet<BlobId>,
+) -> Result<BlobConsistencyCheckResult, TypedStoreError> {
+    // Create a new tokio runtime for the async task to check blob existence.
+    #[cfg(not(msim))]
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .build()
+        .expect("failed to create tokio runtime");
+
     let epoch_bucket = get_epoch_bucket(epoch);
+    let enable_sliver_data_existence_check = node
+        .consistency_check_config
+        .enable_sliver_data_existence_check;
 
     // xxhash is not a cryptographic hash function, but it is fast, has good collision
     // resistance, and is consistent across all platforms.
     let mut hasher = twox_hash::XxHash64::with_seed(epoch as u64);
+    let mut total_synced_scanned = 0;
+    let mut total_fully_stored = 0;
+    let mut existence_check_error = 0;
     for item in blob_info_iter {
         match item {
             Ok(blob_info) => {
                 hasher.write(blob_info.0.as_ref());
+
+                if enable_sliver_data_existence_check
+                    && !blobs_not_yet_fully_synced.contains(&blob_info.0)
+                {
+                    total_synced_scanned += 1;
+
+                    #[cfg(not(msim))]
+                    {
+                        handle_existence_check_result(
+                            // Note that we are running the async function in a new reusable Runtime
+                            // created at the beginning of the task. Note that performance of this
+                            // task is not a concern, and we should limit the resource this task
+                            // can use.
+                            //
+                            // TODO(zhewu): ideally, we should create iterators over the sliver
+                            // column families, and perform sequential scan along with
+                            // BlobInfoIterator to conduct more efficient existence check. This
+                            // requires the SafeIterator to support seek() functionality first.
```

Comment:
This is just unfortunate and I wish we can directly use raw RocksDb interface (which we might, but I didn't dig deeper).

Reply:
This is exactly what I plan to do next. I took a stab at this earlier and the amount of changes to implement this is too much (1000+ line including the change in this PR), so therefore I wanted to do it in a 2 step. That PR itself, namely creating a seek-able iterator, worth its own attending IMO.

Also, given that @sadhansood is importing DBMap trait, I'm thinking now that after he finishes importing it, I can implement it as a function inside DBMap, which makes the seek a native function supported by our version of the DBMap.